### PR TITLE
feat: support for conversion webhooks

### PR DESCRIPTION
**Pull request by diafour**
#### Overview

Use hooks as conversion webhooks.

#### What this PR does / why we need it

See #41. 

This PR adds a new binding `kubernetesCustomResourceConversion`.

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/shell-operator/operator.go [hunk 1]
```diff
@@ -359,13 +363,139 @@ func (op *ShellOperator) InitWebhookManager() (err error) {
...
+		}
+		convPath := op.HookManager.FindConversionChain(event.CrdName, rule)
+		if len(convPath) == 0 {
+			continue
+		}
+		logEntry.Infof("Find conversion path for %s: %v", rule.String(), convPath)
+
+		for _, ruleID := range convPath {
+			var tasks []task.Task
+			op.HookManager.HandleConversionEvent(event, ruleID, func(hook *hook.Hook, info controller.BindingExecutionInfo) {
```
Comment (nabokihms) [inline pkg/shell-operator/operator.go:429]: Can we move the code below to a separate method/function to reduce complexity?

File: pkg/webhook/conversion/chain.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
+package conversion
+
+import (
+	"strings"
+
+	"github.com/flant/shell-operator/pkg/utils/string_helper"
+)
+
+type ChainStorage map[string]*Chain
```
Comment (nabokihms) [inline pkg/webhook/conversion/chain.go:9]: It is better to make ChainStorage a structure with an internal field for the map to make it more flexible.
For example, with the current approach, we can't add a simple lock to protect Chains.

File: pkg/webhook/conversion/chain.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+		for id, path := range newPaths {
+			chain.PathsCache[id] = path
+		}
+	}
+
+	return nil
+}
+
+func (c Chain) SearchPathForRule(rule ConversionRule) []string {
+	IDs := []string{}
```
Comment (nabokihms) [inline pkg/webhook/conversion/chain.go:115]: `ids`

File: pkg/webhook/conversion/config.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+func (r ConversionRule) ShortFromVersion() string {
+	return string_helper.TrimGroup(r.FromVersion)
+}
+
+func (r ConversionRule) ShortToVersion() string {
+	return string_helper.TrimGroup(r.ToVersion)
+}
+
+func RuleFromString(in string) ConversionRule {
```
Comment (nabokihms) [inline pkg/webhook/conversion/config.go:41]: Looks strange. If I got it right, this is our internal format. Why we have to convert rules to strings and vice-versa?

File: pkg/webhook/server/server.go [hunk 5]
```diff
@@ -62,7 +62,7 @@ func (s *WebhookServer) Start() error {
 		tlsConf.ClientCAs = roots
 	}
 
-	listenAddr := app.ValidatingWebhookSettings.ListenAddr + ":" + app.ValidatingWebhookSettings.ListenPort
+	listenAddr := s.Settings.ListenAddr + ":" + s.Settings.ListenPort
```
Comment (nabokihms) [inline pkg/webhook/server/server.go:65]: net package contains precious function `JoinHostPort` for this purpose :)

File: pkg/webhook/server/server.go [hunk 6]
```diff
@@ -85,7 +85,6 @@ func (s *WebhookServer) Start() error {
 		err := srv.ServeTLS(listener, "", "")
 		if err != nil {
 			log.Errorf("Error starting Webhook https server: %v", err)
-			//os.Exit(1)
```
Comment (nabokihms) [inline pkg/webhook/server/server.go:88]: What will happen if the server crashes? I think it is better to restart the shell-operator in this case. There are two ways to do it:
1. Shutdown the shell-operator if one of the webhooks crashes.
2. Provide some sort of health check endpoint to ensure that everything works.

File: pkg/webhook/validating/handler.go [hunk 7]
```diff
@@ -123,20 +123,21 @@ func (h *WebhookHandler) HandleReviewRequest(path string, body []byte) (*v1.Admi
...
-		if configurationId == "" {
-			configurationId = p
+		if configurationID == "" {
+			configurationID = p
 			continue
 		}
 		webhookParts = append(webhookParts, p)
 	}
-	webhookId = strings.Join(webhookParts, "/")
+	webhookID = strings.Join(webhookParts, "/")
```
Comment (nabokihms) [inline pkg/webhook/validating/handler.go:140]: Using `path` is more accurate.

File: examples/204-validating-webhook/README.md [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # Example with validating hooks
 
-This is a simple example of `kubernetesValidating` binding. Read more information in [VALIDATING.md](../../VALIDATING.md).
+This is a simple example of `kubernetesValidating` binding. Read more information in [BINDING_VALIDATING.md](../../BINDING_VALIDATING.md).
```
[the remark this answers is not in the document]
Comment (diafour) [reply examples/204-validating-webhook/README.md:3]: This branch is not rebased.

File: pkg/webhook/validating/handler.go (hunk 7 above)
Comment (diafour) [reply pkg/webhook/validating/handler.go:140]: This webhookID is used as a key later in hook_manager, so the separator string should be constant.

File: pkg/webhook/server/server.go (hunk 6 above)
Comment (diafour) [reply pkg/webhook/server/server.go:88]: Ouch, I need to return this back.

File: pkg/webhook/conversion/config.go (hunk 4 above)
Comment (diafour) [reply pkg/webhook/conversion/config.go:41]: Strings are used to quickly implement a conversion path searching algorithm.

File: pkg/webhook/conversion/chain.go (hunk 2 above)
Comment (diafour) [reply pkg/webhook/conversion/chain.go:9]: Right now lock is not required for ChainStorage — it is filled once at startup. But it should be a good idea to add a lock for maps in Chain structure.

File: pkg/hook/testdata/hook_manager_conversion_chains/hook.sh [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+if [[ $1 == "--config" ]] ; then
+    cat <<EOF
+configVersion: v1
+onStartup: 10
+kubernetesCustomResourceConversion:
+- name: up_conversions
+  crdName: crontabs.stable.example.com
+  conversions:
+  - fromVersion: azaza
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/hook/testdata/hook_manager_conversion_chains/hook.sh:11]: LOL

File: pkg/webhook/validating/handler.go (hunk 7 above)
Comment (nabokihms) [reply pkg/webhook/validating/handler.go:140]: That is exactly what the path package provides.

File: examples/204-validating-webhook/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (nabokihms) [reply examples/204-validating-webhook/README.md:3]: Can you rebase it then?

File: pkg/webhook/conversion/chain.go (hunk 2 above)
Comment (nabokihms) [reply pkg/webhook/conversion/chain.go:9]: A lock is just an example. We can not add new properties to the map, but we add anything we want to the struct.

File: pkg/webhook/validating/handler.go (hunk 7 above)
Comment (diafour) [reply pkg/webhook/validating/handler.go:140]: > // Package path implements utility routines for manipulating slash-separated
// paths.

https://github.com/golang/go/blob/master/src/path/path.go#L5-L6

I thought that `path` uses a system-dependent separator. Anyway, path.Join does a huge redundant [Clean](https://github.com/golang/go/blob/master/src/path/path.go#L70). I would rather use SplitN and drop append+Join.

Comment (nabokihms) [reply pkg/webhook/validating/handler.go:140]: Ok, that sounds reasonable.

File: pkg/shell-operator/operator.go (hunk 1 above)
Comment (nabokihms) [reply pkg/shell-operator/operator.go:429]: We decided to leave it as is.

File: examples/204-validating-webhook/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (diafour) [reply examples/204-validating-webhook/README.md]: Done

File: examples/210-conversion-webhook/templates/certs-secret.yaml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: example-210-conversion-certs
+type: Opaque
```
[the remark this answers is not in the document]
Comment (diafour) [reply examples/210-conversion-webhook/templates/certs-secret.yaml]: Done! Also fixed in 204-validating-webhook example.

File: pkg/webhook/conversion/chain.go (hunk 3 above)
Comment (diafour) [reply pkg/webhook/conversion/chain.go]: Done. No more string ids, Rule struct is used as key.

File: pkg/webhook/conversion/config.go (hunk 4 above)
Comment (diafour) [reply pkg/webhook/conversion/config.go]: Refactored.

File: pkg/webhook/server/server.go (hunk 5 above)
Comment (diafour) [reply pkg/webhook/server/server.go]: Done. JoinHostPort is used now.

File: pkg/webhook/server/server.go (hunk 6 above)
Comment (diafour) [reply pkg/webhook/server/server.go:88]: Reverted!

File: pkg/hook/config/config.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+import (
+	"fmt"
+
+	"sigs.k8s.io/yaml"
+
+	. "github.com/flant/shell-operator/pkg/hook/types"
+)
+
+// HookConfig is a structure with versioned hook configuration
+type HookConfig struct {
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/hook/config/config.go:12]: It also can be "Bindings" or "HookBindings". I'll leave it as is for now until a better naming comes up.

**Comment by diafour** [comment]
Ready: reviewed and tested many times internally.

**Pull request merged**